tools/make.py

```diff
@@ -54,291 +49,343 @@
-        args_error(parser, "one of -p, -n, or --source is required")
-
-    if options.source_dir and not options.build_dir:
-        args_error(parser, "argument --build is required when argument --source is provided")
+        # Target
+        if options.mcu is None:
+            args_error(parser, "argument -m/--mcu is required")
+        mcu = extract_mcus(parser, options)[0]
 
+        # Toolchain
+        if options.tool is None:
+            args_error(parser, "argument -t/--tool is required")
+        toolchain = options.tool[0]
 
-    notify = TerminalNotifier(options.verbose, options.silent, options.color)
+        if (options.program is None) and (not options.source_dir):
+            args_error(parser, "one of -p, -n, or --source is required")
 
-    if not TOOLCHAIN_CLASSES[toolchain].check_executable():
-        search_path = TOOLCHAIN_PATHS[toolchain] or "No path set"
-        args_error(parser, "Could not find executable for %s.\n"
-                           "Currently set search path: %s"
-                           %(toolchain, search_path))
+        if options.source_dir and not options.build_dir:
+            args_error(parser, "argument --build is required when argument --source is provided")
 
-    # Test
-    build_data_blob = {} if options.build_data else None
-    for test_no in p:
-        test = Test(test_no)
-        if options.automated is not None:    test.automated = options.automated
-        if options.dependencies is not None: test.dependencies = options.dependencies
-        if options.host_test is not None:    test.host_test = options.host_test;
-        if options.peripherals is not None:  test.peripherals = options.peripherals;
-        if options.duration is not None:     test.duration = options.duration;
-        if options.extra is not None:        test.extra_files = options.extra
 
-        if not test.is_supported(mcu, toolchain):
-            print('The selected test is not supported on target %s with toolchain %s' % (mcu, toolchain))
-            sys.exit()
+        notify = TerminalNotifier(options.verbose, options.silent, options.color)
 
-        # Linking with extra libraries
-        if options.rpc:      test.dependencies.append(RPC_LIBRARY)
-        if options.usb:      test.dependencies.append(USB_LIBRARIES)
-        if options.dsp:      test.dependencies.append(DSP_LIBRARIES)
-        if options.testlib:  test.dependencies.append(TEST_MBED_LIB)
+        if not TOOLCHAIN_CLASSES[toolchain].check_executable():
+            search_path = TOOLCHAIN_PATHS[toolchain] or "No path set"
+            args_error(parser, "Could not find executable for %s.\n"
+                               "Currently set search path: %s"
+                               %(toolchain, search_path))
 
-        build_dir = join(BUILD_DIR, "test", mcu, toolchain, test.id)
         if options.source_dir is not None:
-            test.source_dir = options.source_dir
-            build_dir = options.source_dir
```

Comment:
I can't seem to find the equivalent of lines 268-270 in the revised `make.py`. It looks like in the old `make.py`, if `--source` was specified and it was building tests, it would reassign the test's source dir as well as the build dir to the specified `--source` directory. Now this old behavior doesn't make much sense to me so I'm fine with the removal, but just wanted to make sure that it was removed for a similar reason.

Reply:
Above, a --source without a --build causes an argument error, so this line was never executed.

Reply:
Good point!